src/packages/cli/src/detach.ts

```diff
@@ -0,0 +1,250 @@
+ */
+export function notifyDetachedInstanceReady() {
+  // in "detach" mode, the parent will wait until the "ready" message is
+  // received before disconnecting from the child process.
+  process.send(READY_MESSAGE);
+}
+
+/**
+ * Attempt to find and remove the instance file for a detached instance.
+ * @param  {string} instanceName the name of the instance to be removed
+ * @returns boolean indicating whether the instance file was cleaned up successfully
+ */
+export async function removeDetachedInstanceFile(
+  instanceName: string
+): Promise<boolean> {
+  const instanceFilename = await getInstanceFilePath(instanceName);
+  try {
+    await rm(instanceFilename);
+    return true;
+  } catch (err) {}
+  return false;
+}
+
+/**
+ * Attempts to stop a detached instance with the specified instance name by
+ * sending a SIGTERM signal. Returns a boolean indicating whether the process
+ * was found. If the PID is identified, but the process is not found, any
+ * corresponding instance file will be removed.
+ *
+ * Note: This does not guarantee that the instance actually stops.
+ * @param  {string} instanceName
+ * @returns boolean indicating whether the instance was found.
+ */
+export async function stopDetachedInstance(
+  instanceName: string
+): Promise<boolean> {
+  try {
+    // getDetachedInstanceByName() throws if the instance file is not found or
+    // cannot be parsed
+    const instance = await getDetachedInstanceByName(instanceName);
+
+    // process.kill() throws if the process was not found (or was a group
+    // process in Windows)
+    process.kill(instance.pid, "SIGTERM");
+  } catch (err) {
+    return false;
+  } finally {
+    removeDetachedInstanceFile(instanceName);
+  }
+  return true;
+}
+
+/**
+ * Start an instance of Ganache in detached mode.
+ * @param  {string[]} argv arguments to be passed to the new instance.
+ * @returns {Promise<DetachedInstance>} resolves to the DetachedInstance once it
+ * is started and ready to receive requests.
+ */
+export async function startDetachedInstance(
+  argv: string[],
+  instanceInfo: {
+    flavor?: FlavorName;
+    server: { host: string; port: number };
+  },
+  version: string
+): Promise<DetachedInstance> {
+  const readingInstanceNames = readdir(await dataPath);
+
+  const [bin, module, ...args] = argv;
+  const childArgs = args.filter(
+    arg => arg !== "--detach" && arg !== "--😈" && arg !== "-D"
+  );
+
+  const child = fork(module, childArgs, {
+    stdio: ["ignore", "ignore", "pipe", "ipc"],
+    detached: true
+  });
+
+  // Any messages output to stderr by the child process (before the `ready`
+  // event is emitted) will be streamed to stderr on the parent.
+  child.stderr.pipe(process.stderr);
+
+  const instanceNames = await readingInstanceNames;
+
+  let instanceName: string;
+  do {
+    instanceName = createInstanceName();
+  } while (instanceNames.indexOf(instanceName) !== -1);
```

Comment:
This is racey. Instead of reading all files in the directory then checking if the random instance name previously existed in the past, we could do _something_ like this completely untested code: 

```
while (true) {
  try {
   const instanceName = createInstanceName();
   const instanceFileName = getInstanceFileName(instanceName);
   details.instanceName = instanceName;
   await writeFile(instanceFilename, details, { flag: "wx" }); // wx means "write, but fail if the file exists"
   break;
  }
  catch(e) {
    if (e.code !== "EEXIST") throw e;
  }
}
```

Reply:
Good find!